File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        
        int n = heights.size();
        vector<int> stack;
        
        int ans = 0;
        
        for(int i = 0; i < n; i ++){
            while(!stack.empty() && (heights[stack.back()] > heights[i])) {
                int top = stack.back();
                stack.pop_back();
                
                int left = -1;
                if(!stack.empty()){
                    left = stack.back();
                }
                
                int length = heights[top];
                int width = i - left - 1;
                
                int area = length * width;
                ans = max(ans, area);
            }
            stack.push_back(i);
        }
        
        return ans;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        
        heights.append(0)
        stack = []
        
        ans = 0
        
        for i in range(len(heights)):
            while stack and heights[stack[-1]] > heights[i]:
                top = stack.pop()
                
                width = i - 1 - (stack[-1] if stack else -1)
                length = heights[top]
                
                area = length *  width
                ans = max(ans, area)
            
            stack.append(i)
        
        return ans